azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunner.java

```diff
@@ -1091,20 +1098,88 @@ public int getExecutionId() {
     return this.execId;
   }
 
+  public Set<JobRunner> getActiveJobRunners() {
+    return ImmutableSet.copyOf(this.activeJobRunners);
+  }
+
+  // Class helps report the flow start and stop events.
+  private class FlowRunnerEventListener implements EventListener {
+
+    public FlowRunnerEventListener() {
+    }
+
+    private synchronized Map<String, String> getFlowMetadata(final FlowRunner flowRunner) {
+      final ExecutableFlow flow = flowRunner.getExecutableFlow();
+      final Map<String, String> metaData = new HashMap<>();
+      metaData.put("flowName", flow.getId());
+      metaData.put("azkabanHost",
+          flowRunner.azkabanProps.getString("azkaban.server.hostname", "unknown"));
+      metaData.put("projectName", flow.getProjectName());
+      metaData.put("submitUser", flow.getSubmitUser());
+      metaData.put("executionId", String.valueOf(flow.getExecutionId()));
+      metaData.put("startTime", String.valueOf(flow.getStartTime()));
+      metaData.put("submitTime", String.valueOf(flow.getSubmitTime()));
+      return metaData;
+    }
+
+    @Override
+    public synchronized void handleEvent(final Event event) {
+      if (FlowRunner.this.azkabanEventReporter == null) {
+        FlowRunner.this.logger.debug("Event reporter is not enabled");
+        return;
+      }
+
+      if (event.getType() == Type.FLOW_STARTED) {
+        final FlowRunner flowRunner = (FlowRunner) event.getRunner();
+        final ExecutableFlow flow = flowRunner.getExecutableFlow();
+        FlowRunner.this.logger.info("Flow started: " + flow.getId());
+        FlowRunner.this.azkabanEventReporter.report(event, getFlowMetadata(flowRunner));
+      } else if (event.getType() == Type.FLOW_FINISHED) {
+        final FlowRunner flowRunner = (FlowRunner) event.getRunner();
+        final ExecutableFlow flow = flowRunner.getExecutableFlow();
+        FlowRunner.this.logger.info("Flow ended: " + flow.getId());
+        final Map<String, String> flowMetadata = getFlowMetadata(flowRunner);
+        flowMetadata.put("endTime", String.valueOf(flow.getEndTime()));
+        FlowRunner.this.azkabanEventReporter.report(event, flowMetadata);
```

Comment:
What happens if the reporter is Null here?

Reply:
Null case is handled at the beginning of the method and it shouldn't reach here if it's null.

Reply:
I see. See the earlier comment on not adding the reporter to the listener list if it is null.

Reply:
done.